quickwit-storage/src/split.rs

```diff
@@ -0,0 +1,469 @@
+use rusoto_core::ByteStream;
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+use tokio_util::io::ReaderStream;
+
+use crate::{BundleStorageFileOffsets, PutPayload};
+
+/// Payload of a split which builds the split bundle and hotcache on the fly and streams it to the
+/// storage.
+#[derive(Clone)]
+pub struct SplitPayload {
+    payloads: Vec<Box<dyn PutPayload>>,
+    /// bytes range of the footer (hotcache + bundle metadata)
+    pub footer_range: Range<u64>,
+}
+
+async fn range_byte_stream_from_payloads(
+    payloads: &[Box<dyn PutPayload>],
+    range: Range<u64>,
+) -> io::Result<ByteStream> {
+    let mut bytestreams: Vec<_> = Vec::new();
+
+    let payloads_and_ranges =
+        chunk_payload_ranges(payloads, range.start as usize..range.end as usize);
+
+    for (payload, range) in payloads_and_ranges {
+        bytestreams.push(
+            payload
+                .range_byte_stream(range.start as u64..range.end as u64)
+                .await?,
+        );
+    }
+
+    let concat_stream = ByteStream::new(stream::iter(bytestreams).flatten());
+    Ok(concat_stream)
+}
+
+#[async_trait]
+impl PutPayload for SplitPayload {
+    fn len(&self) -> u64 {
+        self.payloads.iter().map(|payload| payload.len()).sum()
+    }
+
+    async fn range_byte_stream(&self, range: Range<u64>) -> io::Result<ByteStream> {
+        range_byte_stream_from_payloads(&self.payloads, range).await
+
+        // let requested_payload_ranges = chunk_range(&absolute_payload_ranges, range.start as usize
+        // .. range.end as usize); for payload_range in requested_payload_ranges {
+        // let payload_pos =
+        // absolute_payload_ranges.iter().position(|range|range.contains(&payload_range.start)).
+        // expect("broken");
+
+        // let start=absolute_payload_ranges[payload_pos].start as u64;
+        // let relative_range = payload_range.start as u64 - start .. payload_range.end as u64 -
+        // start;
+
+        // bytestreams.push(self.payloads[payload_pos].range_byte_stream(relative_range).await?);
+
+        //}
+
+        // let mut bytestreams: Vec<_> = Vec::new();
+        // for payload in &self.payloads {
+        // let payload_len = payload.len();
+        // if range.start >= payload.len() {
+        //// The current payload does not intersect with the range
+        // range = (range.start - payload_len)..(range.end - payload_len);
+        // continue;
+        //} else if range.end > payload_len {
+        //// This payload intersects with the range, and it is NOT the last one.
+        // bytestreams.push(payload.range_byte_stream(range.start..payload_len).await?);
+        // range = range.start - payload_len..range.end - payload_len;
+        //} else {
+        //// This is the last chunk
+        // bytestreams.push(payload.range_byte_stream(range.start..range.end).await?);
+        // break;
+        //}
+    }
+}
+
+/// compute split offsets used by the SplitStreamer
+pub fn get_split_payload_streamer(
```

Comment:
with the refactoring and renaming I think `get_split_payload_streamer` is just building the `SplitPayload` and that should be done in the `SplitPayloadBuilder`.